lib/sweet_beats/melody.ex

```diff
@@ -5,17 +5,20 @@ defmodule SweetBeats.Melody do
   @microseconds_between_notes 250_000
 
   def start_link(start_time, instrument \\ Guitar, notes) do
-    GenServer.start_link(__MODULE__, {start_time, instrument, notes})
+    notes_queue = EQueue.from_list(notes)
+    GenServer.start_link(__MODULE__, {start_time, instrument, notes_queue})
   end
 
   def init(state) do
     send(self(), :play)
     {:ok, state}
   end
 
-  def handle_info(:play, {start_time, instrument, [head | tail]}) do
-    play_note(instrument, head)
-    shifted_notes = tail ++ [head]
+  def handle_info(:play, {start_time, instrument, notes_queue}) do
+    {:value, current_note, updated_queue} = EQueue.pop(notes_queue)
+    play_note(instrument, current_note)
+    shifted_notes = EQueue.push(updated_queue, current_note)
```

Comment:
I'd like to extract this into a private `cycle_queue` function which handles the popping and pushing, executing a function with the popped value as it does so, e.g.

```elixir
def handle_info(:play, {start_time, instrument, notes_queue}) do
  updated_queue = 
    cycle_queue(notes_queue, fn(current_note) ->
        play_note(instrument, current_note)
    end)
  # ...
```